README.md
# PlayKit JS KAnalytics - Kaltura Analytics plugin for the [PlayKit JS Player]

[![Build Status](https://travis-ci.org/kaltura/playkit-js-kanalytics.svg?branch=master)](https://travis-ci.org/kaltura/playkit-js-kanalytics)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![](https://img.shields.io/npm/v/@playkit-js/playkit-js-kanalytics/latest.svg)](https://www.npmjs.com/package/@playkit-js/playkit-js-kanalytics)
[![](https://img.shields.io/npm/v/@playkit-js/playkit-js-kanalytics/canary.svg)](https://www.npmjs.com/package/@playkit-js/playkit-js-kanalytics/v/canary)

PlayKit JS KAnalytics plugin integrates Kaltura analytics with the [PlayKit JS Player].

PlayKit JS KAnalytics is written in [ECMAScript6], statically analysed using [Flow] and transpiled in ECMAScript5 using [Babel].

More info about Kaltura Analytics:

- [Video-Analytics]
- [Creating and tracking analytics KMC]

[video-analytics]: https://corp.kaltura.com/Products/Features/Video-Analytics
[creating and tracking analytics kmc]: https://knowledge.kaltura.com/creating-and-tracking-analytics-kmc-0
[flow]: https://flow.org/
[ecmascript6]: https://github.com/ericdouglas/ES6-Learning#articles--tutorials
[babel]: https://babeljs.io

## Getting Started

### Prerequisites

The plugin requires [PlayKit JS Player] to be loaded first.

[playkit js player]: https://github.com/kaltura/playkit-js

### Installing

First, clone and run [yarn] to install dependencies:

[yarn]: https://yarnpkg.com/lang/en/

```
git clone https://github.com/kaltura/playkit-js-kanalytics.git
cd playkit-js-kanalytics
yarn install
```

### Building

Then, build the player

```javascript
yarn run build
```

### Embed the library in your test page

Finally, add the bundle as a script tag in your page, and initialize the player

```html
<script type="text/javascript" src="/PATH/TO/FILE/playkit.js"></script>
<script type="text/javascript" src="/PATH/TO/FILE/playkit-kanalytics.js"></script>
<div id="player-placeholder" style="height:360px; width:640px">
  <script type="text/javascript">
    var playerContainer = document.querySelector("#player-placeholder");
    var config = {
     ...
     plugins: {
       kanalytics: {
         serviceUrl: 'http://stats.kaltura.com/api_v3/index.php'
       }
     }
     ...
    };
    var player = playkit.core.loadPlayer(config);
    playerContainer.appendChild(player.getView());
    player.play();
  </script>
</div>
```

## Configuration

| Settings   | Type   | Required           | Description                |
| ---------- | ------ | ------------------ | -------------------------- |
| serviceUrl | string | :white_check_mark: | The Kaltura API server url |

## Running the tests

Tests can be run locally via [Karma], which will run on Chrome, Firefox and Safari

[karma]: https://karma-runner.github.io/1.0/index.html

```
yarn run test
```

You can test individual browsers:

```
yarn run test:chrome
yarn run test:firefox
yarn run test:safari
```

### And coding style tests

We use ESLint [recommended set](http://eslint.org/docs/rules/) with some additions for enforcing [Flow] types and other rules.

See [ESLint config](.eslintrc.json) for full configuration.

We also use [.editorconfig](.editorconfig) to maintain consistent coding styles and settings, please make sure you comply with the styling.

## Compatibility

TBD

## Contributing

Please read [CONTRIBUTING.md](https://gist.github.com/PurpleBooth/b24679402957c63ec426) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/kaltura/playkit-js-kanalytics/tags).

## License

This project is licensed under the AGPL-3.0 License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.2](https://github.com/kaltura/playkit-js-kanalytics/compare/v1.0.1...v1.0.2) (2022-05-19)



## [1.0.1](https://github.com/kaltura/playkit-js-kanalytics/compare/v1.0.0...v1.0.1) (2020-09-08)



# [1.0.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.11.0...v1.0.0) (2020-08-20)


### Features

* **FEC-10347:** expose kaltura player as a global variable instead of UMD ([#50](https://github.com/kaltura/playkit-js-kanalytics/issues/50)) ([28641f8](https://github.com/kaltura/playkit-js-kanalytics/commit/28641f8a6bda8f7b05acf33d927cbcaa2f541822))


### BREAKING CHANGES

* **FEC-10347:** kanalytics plugin is not UMD anymore 



## [0.11.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.10.1...v0.11.0) (2020-08-05)


### Features

* **FEC-10057:** move the plugin manager to kaltura player ([#44](https://github.com/kaltura/playkit-js-kanalytics/issues/44)) ([f14eb91](https://github.com/kaltura/playkit-js-kanalytics/commit/f14eb91))
* **FEC-10290:** upgrade NPM packages ([#43](https://github.com/kaltura/playkit-js-kanalytics/issues/43)) ([04ac722](https://github.com/kaltura/playkit-js-kanalytics/commit/04ac722))



<a name="0.10.1"></a>
## [0.10.1](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.10.0...v0.10.1) (2019-05-01)



<a name="0.10.0"></a>
# [0.10.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.9.1...v0.10.0) (2018-10-28)


### Features

* **FEC-8452:** playlist ([#37](https://github.com/kaltura/playkit-js-kanalytics/issues/37)) ([2646f27](https://github.com/kaltura/playkit-js-kanalytics/commit/2646f27))



<a name="0.9.1"></a>
## [0.9.1](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.9.0...v0.9.1) (2018-08-22)


### Bug Fixes

* **FEC-8492:** hasKanalony is sent incorrectly ([#36](https://github.com/kaltura/playkit-js-kanalytics/issues/36)) ([a765670](https://github.com/kaltura/playkit-js-kanalytics/commit/a765670))



<a name="0.9.0"></a>
# [0.9.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.8.4...v0.9.0) (2018-06-17)


### Features

* **FEC-8321:** block report if mandatory params are missing ([#32](https://github.com/kaltura/playkit-js-kanalytics/issues/32)) ([edaf560](https://github.com/kaltura/playkit-js-kanalytics/commit/edaf560))



<a name="0.8.4"></a>
## [0.8.4](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.8.3...v0.8.4) (2018-05-26)


### Bug Fixes

* **FEC-8228:** get referrer as config option ([#31](https://github.com/kaltura/playkit-js-kanalytics/issues/31)) ([00b1e58](https://github.com/kaltura/playkit-js-kanalytics/commit/00b1e58))



<a name="0.8.3"></a>
## [0.8.3](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.8.2...v0.8.3) (2018-04-16)


### Bug Fixes

* **FEC-8029:** play reached 100% event sent before ended ([#30](https://github.com/kaltura/playkit-js-kanalytics/issues/30)) ([bb062b9](https://github.com/kaltura/playkit-js-kanalytics/commit/bb062b9))



<a name="0.8.2"></a>
## [0.8.2](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.8.1...v0.8.2) (2018-02-28)


### Bug Fixes

* **FEC-7958:** widget loaded not sent when preload='none' ([#29](https://github.com/kaltura/playkit-js-kanalytics/issues/29)) ([5caa2f9](https://github.com/kaltura/playkit-js-kanalytics/commit/5caa2f9))



<a name="0.8.1"></a>
## [0.8.1](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.7.0...v0.8.1) (2018-02-06)


### Bug Fixes

* **webpack:** playkit core & providers externals root value ([#28](https://github.com/kaltura/playkit-js-kanalytics/issues/28)) ([ccc8e44](https://github.com/kaltura/playkit-js-kanalytics/commit/ccc8e44))


### Features

* update stats service ([#27](https://github.com/kaltura/playkit-js-kanalytics/issues/27)) ([4b7be6a](https://github.com/kaltura/playkit-js-kanalytics/commit/4b7be6a))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.7.0...v0.8.0) (2018-02-06)


### Bug Fixes

* **webpack:** playkit core & providers externals root value ([#28](https://github.com/kaltura/playkit-js-kanalytics/issues/28)) ([ccc8e44](https://github.com/kaltura/playkit-js-kanalytics/commit/ccc8e44))


### Features

* update stats service ([#27](https://github.com/kaltura/playkit-js-kanalytics/issues/27)) ([4b7be6a](https://github.com/kaltura/playkit-js-kanalytics/commit/4b7be6a))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.7.0...v0.8.0) (2018-01-10)


### Features

* update stats service ([#27](https://github.com/kaltura/playkit-js-kanalytics/issues/27)) ([4b7be6a](https://github.com/kaltura/playkit-js-kanalytics/commit/4b7be6a))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.6.3...v0.7.0) (2017-11-16)


### Features

* export name and version on module ([#24](https://github.com/kaltura/playkit-js-kanalytics/issues/24)) ([a6c3019](https://github.com/kaltura/playkit-js-kanalytics/commit/a6c3019))



<a name="0.6.3"></a>
## [0.6.3](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.6.2...v0.6.3) (2017-11-01)


### Bug Fixes

* **FEC-7377:** stats events fail over https ([#22](https://github.com/kaltura/playkit-js-kanalytics/issues/22)) ([e555fb1](https://github.com/kaltura/playkit-js-kanalytics/commit/e555fb1))



<a name="0.6.2"></a>
## [0.6.2](https://github.com/kaltura/playkit-js-kanalytics/compare/v0.6.1...v0.6.2) (2017-10-30)


### Bug Fixes

* **FEC-7357:** send widget loaded event only once ([#19](https://github.com/kaltura/playkit-js-kanalytics/issues/19)) ([579cca5](https://github.com/kaltura/playkit-js-kanalytics/commit/579cca5))
* **FEC-7377:** referrer is missing ([#20](https://github.com/kaltura/playkit-js-kanalytics/issues/20)) ([3fb7949](https://github.com/kaltura/playkit-js-kanalytics/commit/3fb7949))




test/src/kanalytics.spec.js
import '../../src/index';
import {setup} from 'kaltura-player-js';
import * as TestUtils from './utils/test-utils';

describe('KAnalyticsPlugin', function () {
  let player, sandbox, sendSpy, config;
  const playerVersion = '1.2.3';
  const ks =
    'NTAwZjViZWZjY2NjNTRkNGEyMjU1MTg4OGE1NmUwNDljZWJkMzk1MXwxMDY4MjkyOzEwNjgyOTI7MTQ5MDE3NjE0NjswOzE0OTAwODk3NDYuMDIyNjswO3ZpZXc6Kix3aWRnZXQ6MTs7';
  const type = 'vod';
  const id = '1_rwbj3j0a';
  const pId = 1068292;
  const uId = 123456;
  const sId = '7296b4fd-3fcb-666d-51fc-34065579334c';
  const targetId = 'player-placeholder_kanalytics.spec';

  before(function () {
    TestUtils.createElement('DIV', targetId);
    config = {
      targetId,
      provider: {},
      id: id,
      session: {
        partnerID: pId,
        ks: ks,
        uiConfID: uId
      },
      sources: {
        progressive: [
          {
            mimetype: 'video/mp4',
            url: 'https://www.w3schools.com/tags/movie.mp4',
            id: '1_rwbj3j0a_11311,applehttp'
          }
        ]
      },
      plugins: {
        kanalytics: {
          playerVersion: playerVersion,
          entryId: id,
          entryType: type,
          sessionId: sId,
          uiConfId: uId,
          ks: ks,
          partnerId: pId,
          referrer: document.URL
        }
      }
    };
  });

  afterEach(function () {
    sandbox.restore();
    player.destroy();
    TestUtils.removeVideoElementsFromTestPage();
  });

  describe('Basic Playback', function () {
    /**
     * @param {string} ks - ks
     * @param {Object} event - event
     * @return {void}
     */
    function verifyPayloadProperties(payload) {
      payload.ks.should.equal(ks);
      const event = payload.event;
      event.clientVer.should.equal(playerVersion);
      event.partnerId.should.equal(pId);
      event.widgetId.should.equal('_' + pId);
      event.uiConfId.should.equal(uId);
      event.entryId.should.equal(id);
      event.referrer.should.equal(document.URL);
      payload.hasKanalony.should.be.false;
      if (event.duration) {
        event.duration.should.equal(12.612);
      }
      ('playlistId' in event).should.be.false;
    }

    beforeEach(function () {
      sandbox = sinon.createSandbox();
      sendSpy = sandbox.spy(XMLHttpRequest.prototype, 'send');
      player = setup(config);
    });

    it('should send widget loaded before load', () => {
      const payload = sendSpy.firstCall.args[0];
      verifyPayloadProperties(payload);
      payload.event.seek.should.be.false;
      payload.event.eventType.should.equal(1);
    });

    it('should send media loaded', done => {
      player.ready().then(() => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          payload.event.seek.should.be.false;
          payload.event.eventType.should.equal(2);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should send first play', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          payload.event.seek.should.be.false;
          payload.event.eventType.should.equal(3);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.play();
    });

    it('should send replay', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration - 1;
      });
      player.addEventListener(player.Event.ENDED, () => {
        player.addEventListener(player.Event.PLAY, () => {
          try {
            const payload = sendSpy.lastCall.args[0];
            verifyPayloadProperties(payload);
            payload.event.seek.should.be.true;
            payload.event.eventType.should.equal(16);
            done();
          } catch (e) {
            done(e);
          }
        });
        player.play();
      });
      player.play();
    });

    it('should send seek on playing', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration / 2;
      });
      player.addEventListener(player.Event.SEEKED, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          payload.event.seek.should.be.false;
          payload.event.eventType.should.equal(17);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.play();
    });

    it('should send seek before playing', done => {
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.currentTime = player.duration / 2;
      });
      player.addEventListener(player.Event.SEEKED, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          payload.event.seek.should.be.false;
          payload.event.eventType.should.equal(17);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should not send seek for live', done => {
      player._localPlayer._config.sources.type = 'Live';
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration / 2;
      });
      player.addEventListener(player.Event.SEEKED, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          payload.event.eventType.should.not.equal(17);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.play();
    });

    it('should send seek for live + dvr', done => {
      player._localPlayer._config.sources.type = 'Live';
      player._localPlayer._config.sources.dvr = true;
      player.config.sources.type = 'Live';
      player.config.sources.dvr = true;
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration / 2;
      });
      player.addEventListener(player.Event.SEEKED, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          payload.event.eventType.should.equal(17);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.play();
    });
    it('should send buffer start', () => {
      player.dispatchEvent({
        type: player.Event.PLAYER_STATE_CHANGED,
        payload: {
          newState: {
            type: player.State.BUFFERING
          },
          oldState: {
            type: player.State.PLAYING
          }
        }
      });
      const payload = sendSpy.lastCall.args[0];
      verifyPayloadProperties(payload);
      payload.event.seek.should.be.false;
      payload.event.eventType.should.equal(12);
    });

    it('should send buffer end', () => {
      player.dispatchEvent({
        type: player.Event.PLAYER_STATE_CHANGED,
        payload: {
          newState: {
            type: player.State.PLAYING
          },
          oldState: {
            type: player.State.BUFFERING
          }
        }
      });
      const payload = sendSpy.lastCall.args[0];
      verifyPayloadProperties(payload);
      payload.event.seek.should.be.false;
      payload.event.eventType.should.equal(13);
    });

    it('should send 25%', done => {
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.currentTime = 4;
      });
      player.addEventListener(player.Event.TIME_UPDATE, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          if (payload.event.eventType === 4) {
            done();
          }
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should send 50%', done => {
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.currentTime = 7;
      });
      player.addEventListener(player.Event.TIME_UPDATE, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          if (payload.event.eventType === 5) {
            done();
          }
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should send 75%', done => {
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.currentTime = 10;
      });
      player.addEventListener(player.Event.TIME_UPDATE, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          if (payload.event.eventType === 6) {
            done();
          }
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should send 100%', done => {
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.getVideoElement().currentTime = 12.7;
      });
      player.addEventListener(player.Event.TIME_UPDATE, () => {
        try {
          const payload = sendSpy.lastCall.args[0];
          verifyPayloadProperties(payload);
          if (payload.event.eventType === 7) {
            done();
          }
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should send 25% - 100%', done => {
      const onTimeUpdate = () => {
        try {
          player.removeEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
          let payload25 = sendSpy.getCall(1).args[0];
          let payload50 = sendSpy.getCall(2).args[0];
          let payload75 = sendSpy.getCall(3).args[0];
          let payload100 = sendSpy.getCall(4).args[0];
          if (payload25.event.eventType === 2) {
            payload25 = sendSpy.getCall(2).args[0];
            payload50 = sendSpy.getCall(3).args[0];
            payload75 = sendSpy.getCall(4).args[0];
            payload100 = sendSpy.getCall(5).args[0];
          }
          payload25.event.eventType.should.equal(4);
          payload50.event.eventType.should.equal(5);
          payload75.event.eventType.should.equal(6);
          payload100.event.eventType.should.equal(7);
          done();
        } catch (e) {
          done(e);
        }
      };
      player.addEventListener(player.Event.LOADED_METADATA, () => {
        player.addEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
        player.getVideoElement().currentTime = 12.7;
      });
      player.load();
    });

    it('should not send 25% - 100% again while replay', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration - 1;
      });
      player.addEventListener(player.Event.ENDED, () => {
        player.addEventListener(player.Event.PLAY, () => {
          const onTimeUpdate = () => {
            try {
              player.removeEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
              const payload = sendSpy.lastCall.args[0];
              payload.event.eventType.should.equal(16);
              done();
            } catch (e) {
              done(e);
            }
          };
          player.addEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
          player.currentTime = 12.5;
        });
        player.play();
      });
      player.play();
    });
  });

  describe('Change Media', function () {
    const cm_ks =
      'MGI3MzFmMmUwN2IyYmYzN2IxOGEzZjFjMTAzM2U4NTg5MTgyY2MyZnwxMDkxOzEwOTE7MTUwNjY5Mjc0MzswOzE1MDY2MDYzNDMuMTE0MjswO3ZpZXc6Kix3aWRnZXQ6MTs7';
    const cm_type = 'live';
    const cm_id = '1_fdsguh765';
    const cm_plId = '12345678';
    const cm_pId = 2046854;
    const cm_uId = 654321;
    const cm_sId = '15282f1c-fff6-4130-3351-cb8bd39f0cdd';

    const CMconfig = {
      id: cm_id,
      session: {
        partnerID: cm_pId,
        ks: cm_ks,
        uiConfID: cm_uId
      },
      sources: {
        progressive: [
          {
            mimetype: 'video/mp4',
            url: 'https://www.w3schools.com/tags/movie.mp4',
            id: '1_fdsguh765_11311,applehttp'
          }
        ]
      },
      plugins: {
        kanalytics: {
          playerVersion: playerVersion,
          entryId: cm_id,
          playlistId: cm_plId,
          entryType: cm_type,
          sessionId: cm_sId,
          uiConfId: 654321,
          ks: cm_ks,
          partnerId: 2046854,
          referrer: document.URL
        }
      }
    };

    /**
     * @param {Object} event - event
     * @return {void}
     */
    function verifyPayloadProperties(payload) {
      payload.ks.should.equal(cm_ks);
      const event = payload.event;
      event.clientVer.should.equal(playerVersion);
      event.partnerId.should.equal(cm_pId);
      event.widgetId.should.equal('_' + cm_pId);
      event.uiConfId.should.equal(cm_uId);
      event.entryId.should.equal(cm_id);
      event.playlistId.should.equal(cm_plId);
      event.referrer.should.equal(document.URL);
      payload.hasKanalony.should.be.false;
      if (event.duration) {
        event.duration.should.equal(12.612);
      }
    }

    beforeEach(function () {
      sandbox = sinon.createSandbox();
      sendSpy = sandbox.spy(XMLHttpRequest.prototype, 'send');
      player = setup(config);
      player.ready().then(() => {
        player.play();
      });
      player.load();
    });

    it('should not send widget loaded on change media', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_STARTED, () => {
        player.addEventListener(player.Event.SOURCE_SELECTED, () => {
          player.ready().then(() => {
            try {
              const payload = sendSpy.lastCall.args[0];
              verifyPayloadProperties(payload);
              payload.event.seek.should.be.false;
              payload.event.eventType.should.not.equal(1);
              done();
            } catch (e) {
              done(e);
            }
          });
          player.load();
        });
      });
      player.configure(CMconfig);
    });

    it('should send media loaded', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_STARTED, () => {
        player.addEventListener(player.Event.SOURCE_SELECTED, () => {
          player.ready().then(() => {
            try {
              const payload = sendSpy.lastCall.args[0];
              verifyPayloadProperties(payload);
              payload.event.seek.should.be.false;
              payload.event.eventType.should.equal(2);
              done();
            } catch (e) {
              done(e);
            }
          });
          player.load();
        });
      });
      player.configure(CMconfig);
    });

    it('should send first play', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_STARTED, () => {
        player.addEventListener(player.Event.SOURCE_SELECTED, () => {
          player.addEventListener(player.Event.FIRST_PLAY, () => {
            try {
              const payload = sendSpy.lastCall.args[0];
              verifyPayloadProperties(payload);
              payload.event.seek.should.be.false;
              payload.event.eventType.should.equal(3);
              done();
            } catch (e) {
              done(e);
            }
          });
          player.play();
        });
      });
      player.configure(CMconfig);
    });

    it('seek should be false', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_STARTED, () => {
        player.addEventListener(player.Event.SOURCE_SELECTED, () => {
          player.addEventListener(player.Event.FIRST_PLAY, () => {
            try {
              const payload = sendSpy.lastCall.args[0];
              payload.event.seek.should.be.false;
              done();
            } catch (e) {
              done(e);
            }
          });
          player.play();
        });
      });
      const onSeeked = () => {
        player.removeEventListener(player.Event.SEEKED, onSeeked);
        player.configure(CMconfig);
      };
      player.addEventListener(player.Event.SEEKED, onSeeked);
      player.currentTime = 5;
    });

    it('should not send replay if change media on ended', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.currentTime = player.duration - 1;
      });
      player.addEventListener(player.Event.ENDED, () => {
        player.addEventListener(player.Event.LOADED_METADATA, () => {
          player.addEventListener(player.Event.PLAY, () => {
            try {
              const payload = sendSpy.getCall(10).args[0];
              payload.event.eventType.should.not.equal(16);
              done();
            } catch (e) {
              done(e);
            }
          });
          player.play();
        });
        player.configure(CMconfig);
        player.load();
      });
    });

    it('should send 25% - 100% again while change media ', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_STARTED, () => {
        player.addEventListener(player.Event.SOURCE_SELECTED, () => {
          player.addEventListener(player.Event.LOADED_METADATA, () => {
            const onTimeUpdate = () => {
              try {
                player.removeEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
                const payload25 = sendSpy.getCall(9).args[0];
                const payload50 = sendSpy.getCall(10).args[0];
                const payload75 = sendSpy.getCall(11).args[0];
                const payload100 = sendSpy.getCall(12).args[0];
                payload25.event.eventType.should.equal(4);
                payload50.event.eventType.should.equal(5);
                payload75.event.eventType.should.equal(6);
                payload100.event.eventType.should.equal(7);
                verifyPayloadProperties(payload25);
                done();
              } catch (e) {
                done(e);
              }
            };
            player.addEventListener(player.Event.TIME_UPDATE, onTimeUpdate);
            player.getVideoElement().currentTime = 12.7;
          });
          player.load();
        });
      });

      player.addEventListener(player.Event.FIRST_PLAY, () => {
        const onEnded = () => {
          player.removeEventListener(player.Event.TIME_UPDATE, onEnded);
          player.configure(CMconfig);
        };
        player.addEventListener(player.Event.TIME_UPDATE, onEnded);
        player.currentTime = player.duration - 1;
      });
    });
  });

  describe('handle missing params', function () {
    beforeEach(function () {
      sandbox = sinon.createSandbox();
      sendSpy = sandbox.spy(XMLHttpRequest.prototype, 'send');
    });
    it('should not send report if partner id is missing', done => {
      config.plugins.kanalytics.partnerId = '';
      player = setup(config);
      player.ready().then(() => {
        try {
          sendSpy.callCount.should.equal(0);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });

    it('should not send report if entry id is missing', done => {
      config.plugins.kanalytics.entryId = '';
      player = setup(config);
      player.ready().then(() => {
        try {
          sendSpy.callCount.should.equal(0);
          done();
        } catch (e) {
          done(e);
        }
      });
      player.load();
    });
  });
});


src/event.js
//@flow
export default class Event {
  /**
   * The client version
   */
  clientVer: string;
  /**
   * The event type
   */
  eventType: number;
  /**
   * The client's timestamp of this event
   */
  eventTimestamp: number;
  /**
   * The session id. A unique string generated by the client that will represent the client-side session
   * */
  sessionId: string;
  /**
   * The partner id
   */
  partnerId: number;
  /**
   * The entry id
   */
  entryId: string;
  /**
   * The playlist id
   */
  playlistId: ?string;
  /**
   * The widget id
   */
  widgetId: string;
  /**
   * The uiconf id
   */
  uiconfId: number;
  /**
   * The partner's user id
   */
  userId: number;
  /**
   * The timestamp along the video when the event happened
   */
  currentPoint: number;
  /**
   * The duration of the video in milliseconds
   */
  duration: number;
  /**
   * Whether the user ever used seek in this session
   */
  seek: boolean;
  /**
   * The referrer of the client
   */
  referrer: string;
  /**
   * Whether the event is thrown for the first video in the session
   */
  isFirstInSession: boolean;
  /**
   * The kaltura application name
   */
  applicationId: string;
  /**
   * The context id
   */
  contextId: number;
  /**
   * The feature type
   */
  featureType: number;
  /**
   * Whether kanalony plugin is enabled
   */
  hasKanalony: boolean;

  /**
   * @constructor
   * @param {number} eventType - The event type
   */
  constructor(eventType: number) {
    this.eventType = eventType;
    this.isFirstInSession = false;
    this.eventTimestamp = new Date().getTime();
  }
}


karma.conf.js
let webpackConfig = require('./webpack.config.js');
//Need to remove externals otherwise they won't be included in test
delete webpackConfig.externals;
// Need to define inline source maps when using karma
webpackConfig.devtool = 'inline-source-map';

const isWindows = /^win/.test(process.platform);
const isMacOS = /^darwin/.test(process.platform);
// Create custom launcher in case running with Travis
const customLaunchers = {
  Chrome_travis_ci: {
    base: 'Chrome',
    flags: ['--no-sandbox', '--autoplay-policy=no-user-gesture-required']
  }
};

const launchers = {
  Chrome_browser: {
    base: 'Chrome',
    flags: ['--no-sandbox', '--autoplay-policy=no-user-gesture-required']
  }
};

module.exports = function (config) {
  let karmaConf = {
    logLevel: config.LOG_INFO,
    customLaunchers: launchers,
    browsers: ['Chrome_browser', 'Firefox'],
    concurrency: 1,
    singleRun: true,
    colors: true,
    frameworks: ['mocha'],
    files: ['test/setup/karma.js'],
    preprocessors: {
      'src/**/*.js': ['webpack', 'sourcemap'],
      'test/setup/karma.js': ['webpack', 'sourcemap']
    },
    reporters: ['mocha', 'coverage'],
    webpack: webpackConfig,
    webpackServer: {
      noInfo: true
    },
    client: {
      mocha: {
        reporter: 'html',
        timeout: 5000
      }
    }
  };

  if (process.env.TRAVIS) {
    karmaConf.customLaunchers = customLaunchers;
    karmaConf.browsers = ['Chrome_travis_ci'];
  } else {
    if (isWindows) {
      karmaConf.browsers.push('IE');
    } else if (isMacOS) {
      karmaConf.browsers.push('Safari');
    }
  }

  config.set(karmaConf);
};


webpack.config.js
'use strict';

const webpack = require('webpack');
const path = require('path');
const packageData = require('./package.json');

let plugins = [
  new webpack.DefinePlugin({
    __VERSION__: JSON.stringify(packageData.version),
    __NAME__: JSON.stringify(packageData.name)
  })
];

module.exports = {
  context: __dirname + '/src',
  entry: {
    'playkit-kanalytics': 'index.js'
  },
  output: {
    path: __dirname + '/dist',
    filename: '[name].js',
    library: ['playkit', 'kanalytics'],
    devtoolModuleFilenameTemplate: './kanalytics/[resource-path]'
  },
  devtool: 'source-map',
  plugins: plugins,
  module: {
    rules: [
      {
        test: /\.js$/,
        use: [
          {
            loader: 'babel-loader'
          }
        ],
        exclude: [/node_modules/]
      },
      {
        test: /\.js$/,
        exclude: [/node_modules/],
        enforce: 'pre',
        use: [
          {
            loader: 'eslint-loader',
            options: {
              rules: {
                semi: 0
              }
            }
          }
        ]
      }
    ]
  },
  devServer: {
    contentBase: __dirname + '/src'
  },
  resolve: {
    modules: [path.resolve(__dirname, 'src'), 'node_modules']
  },
  externals: {
    'kaltura-player-js': ['KalturaPlayer'],
    'playkit-js-providers': ['KalturaPlayer', 'providers']
  }
};


src/event-types.js
//@flow
const EVENT_TYPES: {[event: string]: number} = {
  WIDGET_LOADED: 1,
  MEDIA_LOADED: 2,
  PLAY: 3,
  PLAY_REACHED_25: 4,
  PLAY_REACHED_50: 5,
  PLAY_REACHED_75: 6,
  PLAY_REACHED_100: 7,
  OPEN_EDIT: 8,
  OPEN_VIRAL: 9,
  OPEN_DOWNLOAD: 10,
  OPEN_REPORT: 11,
  BUFFER_START: 12,
  BUFFER_END: 13,
  OPEN_FULL_SCREEN: 14,
  CLOSE_FULL_SCREEN: 15,
  REPLAY: 16,
  SEEK: 17,
  OPEN_UPLOAD: 18,
  SAVE_PUBLISH: 19,
  CLOSE_EDITOR: 20,
  PRE_BUMPER_PLAYED: 21,
  POST_BUMPER_PLAYED: 22,
  BUMPER_CLICKED: 23,
  PREROLL_STARTED: 24,
  MIDROLL_STARTED: 25,
  POSTROLL_STARTED: 26,
  OVERLAY_STARTED: 27,
  PREROLL_CLICKED: 28,
  MIDROLL_CLICKED: 29,
  POSTROLL_CLICKED: 30,
  OVERLAY_CLICKED: 31,
  PREROLL_25: 32,
  PREROLL_50: 33,
  PREROLL_75: 34,
  MIDROLL_25: 35,
  MIDROLL_50: 36,
  MIDROLL_75: 37,
  POSTROLL_25: 38,
  POSTROLL_50: 39,
  POSTROLL_75: 40
};

export default EVENT_TYPES;
